File: src/mcp_display_image.rs
use std::io::{self, BufRead, Write};

use anyhow::Context;
use serde_json::{json, Value};

const MCP_PROTOCOL_VERSION: &str = "2025-06-18";
const DISPLAY_IMAGE_DESCRIPTION: &str = "Display a workspace-contained image inline in ACP Web UI. Use this after creating, editing, finding, capturing, or referencing an image file that the user should inspect. Prefer this tool over only telling the user the image path.";

pub fn run_stdio() -> anyhow::Result<()> {
    let stdin = io::stdin();
    let mut stdout = io::stdout();

    for line in stdin.lock().lines() {
        let line = line.context("failed to read MCP message")?;
        if line.trim().is_empty() {
            continue;
        }

        let response = match serde_json::from_str::<Value>(&line) {
            Ok(message) => handle_message(&message),
            Err(error) => Some(error_response(Value::Null, -32700, error.to_string())),
        };

        if let Some(response) = response {
            serde_json::to_writer(&mut stdout, &response)?;
            stdout.write_all(b"\n")?;
            stdout.flush()?;
        }
    }

    Ok(())
}

fn handle_message(message: &Value) -> Option<Value> {
    let method = message.get("method").and_then(Value::as_str)?;
    let id = message.get("id").cloned();

    match method {
        "initialize" => id.map(|id| initialize_response(id, message)),
        "tools/list" => id.map(tools_list_response),
        "tools/call" => id.map(|id| tools_call_response(id, message)),
        "ping" => id.map(|id| success_response(id, json!({}))),
        method if method.starts_with("notifications/") => None,
        _ => id.map(|id| error_response(id, -32601, format!("Unsupported MCP method `{method}`"))),
    }
}

fn initialize_response(id: Value, message: &Value) -> Value {
    let requested_version = message
        .get("params")
        .and_then(|params| params.get("protocolVersion"))
        .and_then(Value::as_str)
        .unwrap_or(MCP_PROTOCOL_VERSION);

    success_response(
        id,
        json!({
            "protocolVersion": requested_version,
            "serverInfo": {
                "name": "acp-webui-display-image",
                "version": env!("CARGO_PKG_VERSION")
            },
            "capabilities": {
                "tools": {}
            }
        }),
    )
}

fn tools_list_response(id: Value) -> Value {
    success_response(
        id,
        json!({
            "tools": [
                {
                    "name": "display_image",
                    "description": DISPLAY_IMAGE_DESCRIPTION,
                    "inputSchema": {
                        "type": "object",
                        "properties": {
                            "path": {
                                "type": "string",
                                "description": "Workspace-relative or workspace-contained image path to display inline."
                            },
                            "title": {
                                "type": "string",
                                "description": "Optional concise image title."
                            },
                            "caption": {
                                "type": "string",
                                "description": "Optional short caption for the user."
                            }
                        },
                        "required": ["path"],
                        "additionalProperties": false
                    }
                }
            ]
        }),
    )
}

fn tools_call_response(id: Value, message: &Value) -> Value {
    let params = message.get("params").unwrap_or(&Value::Null);
    let name = params
        .get("name")
        .and_then(Value::as_str)
        .unwrap_or_default();
    if name != "display_image" {
        return success_response(
            id,
            json!({
                "content": [
                    {
                        "type": "text",
                        "text": format!("Unknown tool `{name}`")
                    }
                ],
                "isError": true
            }),
        );
    }

    let arguments = params.get("arguments").unwrap_or(&Value::Null);
    let Some(path) = arguments
        .get("path")
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|path| !path.is_empty())
    else {
        return success_response(
            id,
            json!({
                "content": [
                    {
                        "type": "text",
                        "text": "display_image requires a non-empty path argument."
                    }
                ],
                "isError": true
            }),
        );
    };

    success_response(
        id,
        json!({
            "content": [
                {
                    "type": "text",
                    "text": format!("display_image requested inline display for image `{path}`.")
                }
            ],
            "structuredContent": {
                "path": path,
                "title": arguments.get("title").and_then(Value::as_str),
                "caption": arguments.get("caption").and_then(Value::as_str)
            }
        }),
    )
}

fn success_response(id: Value, result: Value) -> Value {
    json!({
        "jsonrpc": "2.0",
        "id": id,
        "result": result
    })
}

fn error_response(id: Value, code: i64, message: String) -> Value {
    json!({
        "jsonrpc": "2.0",
        "id": id,
        "error": {
            "code": code,
            "message": message
        }
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lists_display_image_tool() {
        let response = handle_message(&json!({
            "jsonrpc": "2.0",
            "id": 1,
            "method": "tools/list"
        }))
        .unwrap();

        assert_eq!(
            response["result"]["tools"][0]["name"],
            Value::String("display_image".to_string())
        );
    }

    #[test]
    fn display_image_tool_requires_path() {
        let response = handle_message(&json!({
            "jsonrpc": "2.0",
            "id": 1,
            "method": "tools/call",
            "params": {
                "name": "display_image",
                "arguments": {}
            }
        }))
        .unwrap();

        assert_eq!(response["result"]["isError"], Value::Bool(true));
    }

    #[test]
    fn display_image_tool_returns_structured_path() {
        let response = handle_message(&json!({
            "jsonrpc": "2.0",
            "id": 1,
            "method": "tools/call",
            "params": {
                "name": "display_image",
                "arguments": {
                    "path": "prototype/screenshots/overview.png"
                }
            }
        }))
        .unwrap();

        assert_eq!(
            response["result"]["structuredContent"]["path"],
            Value::String("prototype/screenshots/overview.png".to_string())
        );
    }
}

File: mcp_display_image.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"strings"
)

const mcpProtocolVersion = "2025-06-18"

const displayImageDescription = "Display a workspace-contained image inline in ACP Web UI. Use this after creating, editing, finding, capturing, or referencing an image file that the user should inspect. Prefer this tool over only telling the user the image path."

func runMCPDisplayImage(input io.Reader, output io.Writer) error {
	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var message map[string]any
		var response any
		if err := json.Unmarshal([]byte(line), &message); err != nil {
			response = mcpError(nil, -32700, err.Error())
		} else {
			response = handleMCPMessage(message)
		}
		if response == nil {
			continue
		}
		data, _ := json.Marshal(response)
		if _, err := output.Write(append(data, '\n')); err != nil {
			return err
		}
	}
	return scanner.Err()
}

func handleMCPMessage(message map[string]any) any {
	method, _ := message["method"].(string)
	id, hasID := message["id"]
	if !hasID {
		return nil
	}
	switch {
	case method == "initialize":
		version := mcpProtocolVersion
		if params, ok := message["params"].(map[string]any); ok {
			if requested, ok := params["protocolVersion"].(string); ok && requested != "" {
				version = requested
			}
		}
		return mcpSuccess(id, map[string]any{
			"protocolVersion": version,
			"serverInfo":      map[string]any{"name": "acp-webui-display-image", "version": "0.1.0"},
			"capabilities":    map[string]any{"tools": map[string]any{}},
		})
	case method == "tools/list":
		return mcpSuccess(id, map[string]any{"tools": []any{map[string]any{
			"name":        "display_image",
			"description": displayImageDescription,
			"inputSchema": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"path":    map[string]any{"type": "string", "description": "Workspace-relative or workspace-contained image path to display inline."},
					"title":   map[string]any{"type": "string", "description": "Optional concise image title."},
					"caption": map[string]any{"type": "string", "description": "Optional short caption for the user."},
				},
				"required":             []string{"path"},
				"additionalProperties": false,
			},
		}}})
	case method == "tools/call":
		return mcpToolCall(id, message)
	case method == "ping":
		return mcpSuccess(id, map[string]any{})
	case strings.HasPrefix(method, "notifications/"):
		return nil
	default:
		return mcpError(id, -32601, fmt.Sprintf("Unsupported MCP method `%s`", method))
	}
}

func mcpToolCall(id any, message map[string]any) any {
	params, _ := message["params"].(map[string]any)
	name, _ := params["name"].(string)
	if name != "display_image" {
		return mcpSuccess(id, map[string]any{
			"content": []any{map[string]any{"type": "text", "text": fmt.Sprintf("Unknown tool `%s`", name)}},
			"isError": true,
		})
	}
	args, _ := params["arguments"].(map[string]any)
	path, _ := args["path"].(string)
	path = strings.TrimSpace(path)
	if path == "" {
		return mcpSuccess(id, map[string]any{
			"content": []any{map[string]any{"type": "text", "text": "display_image requires a non-empty path argument."}},
			"isError": true,
		})
	}
	return mcpSuccess(id, map[string]any{
		"content": []any{map[string]any{"type": "text", "text": fmt.Sprintf("display_image requested inline display for image `%s`.", path)}},
		"structuredContent": map[string]any{
			"path":    path,
			"title":   args["title"],
			"caption": args["caption"],
		},
	})
}

func mcpSuccess(id any, result any) map[string]any {
	return map[string]any{"jsonrpc": "2.0", "id": id, "result": result}
}

func mcpError(id any, code int, message string) map[string]any {
	return map[string]any{"jsonrpc": "2.0", "id": id, "error": map[string]any{"code": code, "message": message}}
}
